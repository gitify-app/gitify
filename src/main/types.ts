export interface GitifyAPI {
  openExternalLink: (url: string) => void;
  getAppVersion: () => Promise<string>;
  encryptValue: (value: string) => Promise<string>;
  decryptValue: (value: string) => Promise<string>;
  quitApp: () => void;
  showWindow: () => void;
  hideWindow: () => void;
  setAutoLaunch: (value: boolean) => void;
  setAlternateIdleIcon: (value: boolean) => void;
  setKeyboardShortcut: (keyboardShortcut: boolean) => void;
  updateTrayIcon: (notificationsLength?: number) => void;
  updateTrayTitle: (title?: string) => void;
}
