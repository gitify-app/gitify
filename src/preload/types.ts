import type { api } from '.';

export type GitifyAPI = typeof api;
