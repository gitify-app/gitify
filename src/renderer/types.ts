import type { OcticonProps } from '@primer/octicons-react';
import type { FC } from 'react';
import type { Notification, Reason } from './typesGitHub';
import type {
  AuthMethod,
  EnterpriseAccount,
  PlatformType,
} from './utils/auth/types';

declare const __brand: unique symbol;

type Brand<B> = { [__brand]: B };

export interface AuthState {
  accounts: Account[];
  /**
   * @deprecated This attribute is deprecated and will be removed in a future release.
   */
  token?: Token;
  /**
   * @deprecated This attribute is deprecated and will be removed in a future release.
   */
  enterpriseAccounts?: EnterpriseAccount[];
  /**
   * @deprecated This attribute is deprecated and will be removed in a future release.
   */
  user?: GitifyUser | null;
}

export type Branded<T, B> = T & Brand<B>;

export type AuthCode = Branded<string, 'AuthCode'>;

export type Token = Branded<string, 'Token'>;

export type ClientID = Branded<string, 'ClientID'>;

export type ClientSecret = Branded<string, 'ClientSecret'>;

export type Hostname = Branded<string, 'Hostname'>;

export type Link = Branded<string, 'WebUrl'>;

export type Status = 'loading' | 'success' | 'error';

export interface Account {
  method: AuthMethod;
  platform: PlatformType;
  version?: string;
  hostname: Hostname;
  token: Token;
  user: GitifyUser | null;
  hasRequiredScopes?: boolean;
}

export type SettingsValue =
  | boolean
  | number
  | GroupBy
  | OpenPreference
  | Reason[]
  | Theme;

export type SettingsState = AppearanceSettingsState &
  NotificationSettingsState &
  SystemSettingsState &
  FilterSettingsState;

export interface AppearanceSettingsState {
  theme: Theme;
  zoomPercentage: number;
  detailedNotifications: boolean;
  showAccountHeader: boolean;
  showPills: boolean;
  showNumber: boolean;
  wrapNotificationTitle: boolean;
}

export interface NotificationSettingsState {
  groupBy: GroupBy;
  fetchAllNotifications: boolean;
  participating: boolean;
  markAsDoneOnOpen: boolean;
  markAsDoneOnUnsubscribe: boolean;
  delayNotificationState: boolean;
}

export interface SystemSettingsState {
  openLinks: OpenPreference;
  keyboardShortcut: boolean;
  showNotificationsCountInTray: boolean;
  showNotifications: boolean;
  useAlternateIdleIcon: boolean;
  playSound: boolean;
  openAtStartup: boolean;
}

export interface FilterSettingsState {
  hideBots: boolean;
  filterReasons: Reason[];
}

export interface GitifyState {
  auth?: AuthState;
  settings?: SettingsState;
}

export enum Theme {
  SYSTEM = 'SYSTEM',
  LIGHT = 'LIGHT',
  LIGHT_HIGH_CONTRAST = 'LIGHT_HIGH_CONTRAST',
  LIGHT_COLORBLIND = 'LIGHT_COLORBLIND',
  LIGHT_TRITANOPIA = 'LIGHT_TRITANOPIA',
  DARK = 'DARK',
  DARK_HIGH_CONTRAST = 'DARK_HIGH_CONTRAST',
  DARK_COLORBLIND = 'DARK_COLORBLIND',
  DARK_TRITANOPIA = 'DARK_TRITANOPIA',
  DARK_DIMMED = 'DARK_DIMMED',
}

export enum OpenPreference {
  FOREGROUND = 'FOREGROUND',
  BACKGROUND = 'BACKGROUND',
}

export enum GroupBy {
  REPOSITORY = 'REPOSITORY',
  DATE = 'DATE',
}

export type RadioGroupItem = {
  label: string;
  value: string;
};

export interface AccountNotifications {
  account: Account;
  notifications: Notification[];
  error: GitifyError | null;
}

export interface GitifyUser {
  login: string;
  name: string | null;
  avatar: Link | null;
  id: number;
}

export interface GitifyError {
  title: string;
  descriptions: string[];
  emojis: string[];
}

export type ErrorType =
  | 'BAD_CREDENTIALS'
  | 'MISSING_SCOPES'
  | 'NETWORK'
  | 'RATE_LIMITED'
  | 'UNKNOWN';

export interface FormattedReason {
  title: string;
  description: string;
}

export enum IconColor {
  GRAY = 'text-gitify-icon-muted',
  GREEN = 'text-gitify-icon-open',
  PURPLE = 'text-gitify-icon-done',
  RED = 'text-gitify-icon-closed',
  YELLOW = 'text-gitify-icon-attention',
}

export enum Opacity {
  READ = 'opacity-50',
  LOW = 'opacity-70',
  MEDIUM = 'opacity-80',
  HIGH = 'opacity-90',
}

export type PullRequestApprovalIcon = {
  type: FC<OcticonProps>;
  color: IconColor;
  description: string;
};

export enum Size {
  XSMALL = 12,
  SMALL = 14,
  MEDIUM = 16,
  LARGE = 18,
  XLARGE = 20,
}

export type Chevron = {
  icon: FC<OcticonProps>;
  label: string;
};
