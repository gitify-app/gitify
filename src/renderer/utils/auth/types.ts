import type {
  AuthCode,
  ClientID,
  ClientSecret,
  Hostname,
  Token,
} from '../../types';

export type AuthMethod = 'GitHub App' | 'Personal Access Token' | 'OAuth App';

export type PlatformType = 'GitHub Cloud' | 'GitHub Enterprise Server';

export interface LoginOAuthAppOptions {
  hostname: Hostname;
  clientId: ClientID;
  clientSecret: ClientSecret;
}

export interface LoginPersonalAccessTokenOptions {
  hostname: Hostname;
  token: Token;
}

export interface AuthResponse {
  authMethod: AuthMethod;
  authCode: AuthCode;
  authOptions: LoginOAuthAppOptions;
}

export interface AuthTokenResponse {
  hostname: Hostname;
  token: Token;
}
