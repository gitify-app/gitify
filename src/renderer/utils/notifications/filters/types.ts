import type {
  AccountNotifications,
  SettingsState,
  TypeDetails,
} from '../../../types';
import type { Notification } from '../../../typesGitHub';

export interface Filter<T extends string> {
  FILTER_TYPES: Record<T, TypeDetails>;

  requiresDetailsNotifications: boolean;

  getTypeDetails(type: T): TypeDetails;

  hasFilters(settings: SettingsState): boolean;

  isFilterSet(settings: SettingsState, type: T): boolean;

  getFilterCount(notifications: AccountNotifications[], type: T): number;

  filterNotification(notification: Notification, type: T): boolean;
}
