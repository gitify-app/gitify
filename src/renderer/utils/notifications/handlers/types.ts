import type { FC } from 'react';

import type { OcticonProps } from '@primer/octicons-react';

import type { SettingsState } from '../../../types';
import type {
  GitifySubject,
  Notification,
  Subject,
  SubjectType,
} from '../../../typesGitHub';

export interface NotificationTypeHandler {
  readonly type?: SubjectType;

  /**
   * Enrich a notification. Settings may be unused for some handlers.
   */
  enrich(
    notification: Notification,
    settings: SettingsState,
  ): Promise<GitifySubject>;

  /** Return an icon component for this notification type. */
  getIcon(subject: Subject): FC<OcticonProps> | null;
}
