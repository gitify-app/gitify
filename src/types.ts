import type { Notification } from './typesGithub';

export interface AuthState {
  token?: string;
  enterpriseAccounts: EnterpriseAccount[];
  user: GitifyUser | null;
}

export type SettingsState = AppearanceSettingsState &
  NotificationSettingsState &
  SystemSettingsState;

interface AppearanceSettingsState {
  theme: Theme;
  detailedNotifications: boolean;
  showAccountHostname: boolean;
}

interface NotificationSettingsState {
  participating: boolean;
  showNotifications: boolean;
  showBots: boolean;
  markAsDoneOnOpen: boolean;
}

interface SystemSettingsState {
  playSound: boolean;
  openAtStartup: boolean;
  showNotificationsCountInTray: boolean;
}

export enum Theme {
  SYSTEM = 'SYSTEM',
  LIGHT = 'LIGHT',
  DARK = 'DARK',
}

export type RadioGroupItem = {
  label: string;
  value: string;
};

export interface EnterpriseAccount {
  hostname: string;
  token: string;
}

export interface AccountNotifications {
  hostname: string;
  notifications: Notification[];
}

export interface AuthOptions {
  hostname: string;
  clientId: string;
  clientSecret: string;
}

export interface AuthTokenOptions {
  hostname: string;
  token: string;
}

export interface AuthResponse {
  authCode: string;
  authOptions: AuthOptions;
}
export interface AuthTokenResponse {
  hostname: string;
  token: string;
}

export interface GitifyUser {
  login: string;
  name: string;
  id: number;
}

export interface GitifyError {
  title: string;
  descriptions: string[];
  emojis: string[];
}

export type ErrorType =
  | 'BAD_CREDENTIALS'
  | 'MISSING_SCOPES'
  | 'NETWORK'
  | 'RATE_LIMITED'
  | 'UNKNOWN';

export interface FormattedReason {
  title: string;
  description: string;
}
