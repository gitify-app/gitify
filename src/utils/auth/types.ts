import type {
  AuthCode,
  ClientID,
  ClientSecret,
  Hostname,
  Token,
} from '../../types';

export type AuthMethod = 'GitHub App' | 'Personal Access Token' | 'OAuth App';

export type PlatformType = 'GitHub Cloud' | 'GitHub Enterprise Server';

export interface LoginOAuthAppOptions {
  hostname: Hostname;
  clientId: ClientID;
  clientSecret: ClientSecret;
}

export interface LoginPersonalAccessTokenOptions {
  hostname: Hostname;
  token: Token;
}

export interface AuthResponse {
  authCode: AuthCode;
  authOptions: LoginOAuthAppOptions;
}

export interface AuthTokenResponse {
  hostname: Hostname;
  token: Token;
}

/**
 * @deprecated This type is deprecated and will be removed in a future release.
 */
export interface EnterpriseAccount {
  hostname: Hostname;
  token: Token;
}
