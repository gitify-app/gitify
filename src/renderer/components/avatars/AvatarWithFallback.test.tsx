import { render, screen } from '@testing-library/react';

import { type Link, Size } from '../../types';
import {
  AvatarWithFallback,
  type IAvatarWithFallback,
} from './AvatarWithFallback';

describe('renderer/components/avatars/AvatarWithFallback.tsx', () => {
  const props: IAvatarWithFallback = {
    src: 'https://avatars.githubusercontent.com/u/133795385?s=200&v=4' as Link,
    alt: 'gitify-app',
    name: '@gitify-app',
    size: Size.MEDIUM,
    userType: 'User',
  };

  it('should render avatar - human user', () => {
    const tree = render(<AvatarWithFallback {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render avatar - non-human user', () => {
    const tree = render(
      <AvatarWithFallback {...props} userType={'Organization'} />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders the fallback icon when no src url - human user', () => {
    const tree = render(<AvatarWithFallback {...props} src={null} />);

    expect(tree).toMatchSnapshot();
  });

  it('renders the fallback icon when no src url - non human user', () => {
    const tree = render(
      <AvatarWithFallback {...props} src={null} userType="Bot" />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders the fallback icon when the image fails to load (isBroken = true) - human user', () => {
    render(<AvatarWithFallback {...props} />);

    // Find the avatar element by its alt text
    const avatar = screen.getByAltText('gitify-app') as HTMLImageElement;

    // Simulate an error event on the image element
    avatar.dispatchEvent(new Event('error'));

    expect(screen.getByTestId('avatar')).toMatchSnapshot();
  });

  it('renders the fallback icon when the image fails to load (isBroken = true) - non human user', () => {
    render(<AvatarWithFallback {...props} userType={'Bot'} />);

    // Find the avatar element by its alt text
    const avatar = screen.getByAltText('gitify-app') as HTMLImageElement;

    // Simulate an error event on the image element
    avatar.dispatchEvent(new Event('error'));

    expect(screen.getByTestId('avatar')).toMatchSnapshot();
  });
});
