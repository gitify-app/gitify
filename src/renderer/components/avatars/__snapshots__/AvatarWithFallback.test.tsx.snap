// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer/components/avatars/AvatarWithFallback.tsx renders the fallback icon when the image fails to load (isBroken = true) - human user 1`] = `
<div
  class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
  data-align="center"
  data-direction="horizontal"
  data-gap="condensed"
  data-justify="start"
  data-padding="none"
  data-testid="avatar"
  data-wrap="nowrap"
>
  <img
    alt="gitify-app"
    class="Avatar__StyledAvatar-sc-2lv0r8-0 kuZlm"
    data-component="Avatar"
    height="16"
    size="16"
    src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
    sx="[object Object]"
    width="16"
  />
  <span
    class="Text-sc-17v1xeu-0 hmAFGJ"
  >
    @gitify-app
  </span>
</div>
`;

exports[`renderer/components/avatars/AvatarWithFallback.tsx renders the fallback icon when the image fails to load (isBroken = true) - non human user 1`] = `
<div
  class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
  data-align="center"
  data-direction="horizontal"
  data-gap="condensed"
  data-justify="start"
  data-padding="none"
  data-testid="avatar"
  data-wrap="nowrap"
>
  <img
    alt="gitify-app"
    class="Avatar__StyledAvatar-sc-2lv0r8-0 fdrsHc"
    data-component="Avatar"
    height="16"
    size="16"
    src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
    sx="[object Object]"
    width="16"
  />
  <span
    class="Text-sc-17v1xeu-0 hmAFGJ"
  >
    @gitify-app
  </span>
</div>
`;

exports[`renderer/components/avatars/AvatarWithFallback.tsx should render avatar - human user 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
        data-align="center"
        data-direction="horizontal"
        data-gap="condensed"
        data-justify="start"
        data-padding="none"
        data-testid="avatar"
        data-wrap="nowrap"
      >
        <img
          alt="gitify-app"
          class="Avatar__StyledAvatar-sc-2lv0r8-0 kuZlm"
          data-component="Avatar"
          height="16"
          size="16"
          src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
          sx="[object Object]"
          width="16"
        />
        <span
          class="Text-sc-17v1xeu-0 hmAFGJ"
        >
          @gitify-app
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
      data-align="center"
      data-direction="horizontal"
      data-gap="condensed"
      data-justify="start"
      data-padding="none"
      data-testid="avatar"
      data-wrap="nowrap"
    >
      <img
        alt="gitify-app"
        class="Avatar__StyledAvatar-sc-2lv0r8-0 kuZlm"
        data-component="Avatar"
        height="16"
        size="16"
        src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
        sx="[object Object]"
        width="16"
      />
      <span
        class="Text-sc-17v1xeu-0 hmAFGJ"
      >
        @gitify-app
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`renderer/components/avatars/AvatarWithFallback.tsx should render avatar - non-human user 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
        data-align="center"
        data-direction="horizontal"
        data-gap="condensed"
        data-justify="start"
        data-padding="none"
        data-testid="avatar"
        data-wrap="nowrap"
      >
        <img
          alt="gitify-app"
          class="Avatar__StyledAvatar-sc-2lv0r8-0 fdrsHc"
          data-component="Avatar"
          height="16"
          size="16"
          src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
          sx="[object Object]"
          width="16"
        />
        <span
          class="Text-sc-17v1xeu-0 hmAFGJ"
        >
          @gitify-app
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
      data-align="center"
      data-direction="horizontal"
      data-gap="condensed"
      data-justify="start"
      data-padding="none"
      data-testid="avatar"
      data-wrap="nowrap"
    >
      <img
        alt="gitify-app"
        class="Avatar__StyledAvatar-sc-2lv0r8-0 fdrsHc"
        data-component="Avatar"
        height="16"
        size="16"
        src="https://avatars.githubusercontent.com/u/133795385?s=200&v=4"
        sx="[object Object]"
        width="16"
      />
      <span
        class="Text-sc-17v1xeu-0 hmAFGJ"
      >
        @gitify-app
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
